Python bindings

shakamd opened the pull request:
Talk to the Sia network with python! Currently requires a siad instance, but could easily support shard+walrus.

lukechampine commented:
Ah, this is great, the name is even a pun (intentional or not)!

I'll definitely merge this, but two quick questions:

1. Seems a shame to duplicate the C bindings in two places. Is it possible to reuse the existing bindings, or is there Python-specific stuff in yours?
2. Could we (optionally) use `with` in the Python API for files, like the blocks in the Ruby API?

shakamd commented:
The main differences in the bindings are the addition of some lower level "us_ll" wrappers to access some of the contract and session functions (and likely more in future). The other difference is the addition of an 'id' parameter to each of the wrapper functions. The id is the instance address of the python caller class, and it's there to uniquely index us_err (which I changed in to a map) so that various instances of pyus objects do not clobber each other's error state.

Good point on context managers, I've refactored the code and the filesystem example to more closely match the ruby example.

lukechampine commented on python/bindings.go:
```diff
@@ -0,0 +1,397 @@
+    copy(sector[:], goBytes(buf, renterhost.SectorSize))
+    err := session.Write([]renterhost.RPCWriteAction{{
+              Type: renterhost.RPCWriteActionAppend,
+              Data: sector[:],
```
ooh, I actually [added a helper method for this](https://github.com/lukechampine/us/commit/f3ce7ddc816e1533dc723a1e1592156424caaf13) not too long ago:
```go
root, err := session.Append(&sector)
```
More convenient to use, and faster too, because it computes the sector root in parallel :)

lukechampine replied:
I guess this raises the question of versioning: how do we ensure that this library is compiled with at least `us@v0.12.0`?

shakamd replied:
Can we not just include a go.mod file with something like
```
module lukechampine.com/us-bindings/python

go 1.13

require lukechampine.com/us v0.14.0
```

lukechampine replied:
Probably. I'll look into it.

lukechampine commented on python/bindings.go:
```diff
@@ -0,0 +1,397 @@
+    copy(sectorMerkleRoot[:], goBytes(root, crypto.HashSize))
+    var b bytes.Buffer
+    dst := bufio.NewWriter(&b)
+    err := session.Read(dst, []renterhost.RPCReadRequestSection{{
```
I think the `bufio` is actually unnecessary here. You'd want to use `bufio` when `Write` is an expensive operation, but here we're just writing into a `bytes.Buffer`, which should be very fast.

Also, probably best to call `b.Grow(int(length))` on the buffer so that we don't need to re-allocate it.

shakamd replied:
Is there a way to massage `bytes.Buffer` into an `io.Writer` type? That's what the session Read call expects, `bufio` made it stop complaining.

lukechampine replied:
Ah, you just need to pass a pointer to the buffer. Alternatively, you could use `bytes.NewBuffer`, which returns a pointer and lets you pre-allocate up front:
```go
b := bytes.NewBuffer(make([]byte, 0, length))
err := session.Read(&b, ...
```

lukechampine commented on python/example/filesystem.py:
```diff
@@ -0,0 +1,39 @@
+
+with pyus.FileSystem("meta", hs) as fs:
+    # create a file
+    with fs.create("foo.txt", 1) as f:
```
👍

lukechampine commented on python/bindings.go:
```diff
@@ -0,0 +1,389 @@
+    session := loadPtr(session_p).(*proto.Session)
+    var sector [renterhost.SectorSize]byte
+    copy(sector[:], goBytes(buf, renterhost.SectorSize))
+    root, err := session.Append(&sector)
```
you could probably save a copy here by casting `buf` to a  `*[renterhost.SectorSize]byte` directly. That's a little dangerous, though. You want to have some protection against the caller passing a "sector" with the wrong size.

Fine for now, though.
